lib/ruby_llm/providers/ollama/tools.rb

```diff
@@ -48,6 +48,26 @@ def parse_tool_calls(tool_calls)
           end.compact
         end
 
+        # HACK: Llama3.x yields this in the proper tool_calls response field,
+        # but some other models return it as markup inside the text response.
+        #
+        # Unfortunately said other models are all over the place when it comes
+        # to sticking to a format so this doesn't cover all edge cases.
+        def preprocess_tool_calls(response_data)
+          # Move <toolcall>JSON</toolcall> markup from inside the text to its
+          # specific field in the response where it should be.
+          # https://github.com/ollama/ollama/blob/main/docs/api.md#chat-request-with-tools
+
```

Comment:
This is a bit of an abomination to attempt to mitigate the problems of attempting tool calling on smaller models; both their training and Ollama support is a bit undercooked.

Llama 3 does not seem to require this, but others do and it would be great to still support them. I could add a unit test excercise the hairy logic.

Reply:
Unit test added: https://github.com/crmne/ruby_llm/pull/10/commits/a3635b5fc614f09832cb61b1090d8e236012a2e4